### PR TITLE
[ModalSheet] Add data based ModalSheet for dynamic content

### LINKED ISSUE #1: [ModalSheet] No exit animation when ModalSheet has dynamic content
## Description
Sometimes it is needed to show dynamic data in modal sheet content. Imagine error bottom sheet use-case, where if `Error` is not null we need to show modal, otherwise hide it. We could either wrap `ModalSheet` in "if" condition to show only for non-null error, or adjust content (which is animated) for null error. 

Both possible solutions did not provide good experience. Either there is no exit animation, or content is "blinking" when animating out.

## Sample
```

data class ErrorData(val title: String, val message: String)

val error = rememberSaveable { mutableStateOf<ErrorData?>(null) }

Box(
    contentAlignment = Alignment.Center,
    modifier = Modifier
        .fillMaxSize()
        .padding(padding)
) {
    Button(onClick = {
        error.value = ErrorData(
            title = "Error",
            message = "Oops, an error occurred"
        )
    }) {
        Text(text = "Error!")
    }
}

// Solution #1 - wrap in if condition
val errorValue = error.value
if (errorValue != null) {
    ModalSheet(
        visible = true,
        onDismiss = { error.value = null }
    ) {
        Column(
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(SpacingSmall),
            modifier = Modifier
                .fillMaxWidth()
                .navigationBarsPadding()
                .padding(SpacingMedium)
        ) {
            Text(
                text = errorValue.title,
                style = MaterialTheme.typography.h4
            )
            Text(
                text = errorValue.message,
            )
        }
    }
}

// Solution #2 - null content is visible when animated out
ModalSheet(
    visible = error.value != null,
    onDismiss = { error.value = null }
) {
    Column(
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(SpacingSmall),
        modifier = Modifier
            .fillMaxWidth()
            .navigationBarsPadding()
            .padding(SpacingMedium)
    ) {
        Text(
            text = error.value?.title.orEmpty(),
            style = MaterialTheme.typography.h4
        )
        Text(
            text = error.value?.message.orEmpty(),
        )
    }
}
```

## Video
First solution to wrap in if condition, results in no exit animation.
https://user-images.githubusercontent.com/20944869/167183075-409de54f-6954-4206-bc90-b488755e6958.mp4

Second solution to use null in content, results in "empty" content during exit animation.
https://user-images.githubusercontent.com/20944869/167187758-c1d16bb1-d21e-44ec-adf1-78bf76763c3a.mp4

### PR DESCRIPTION
**Pull request by oleksandrbalan**
Should fix #1 

Do not really like the API though :/ 

@wooodenleg What do you think?

```
ModalSheet(
    data = error,
    visible = { this != null },
    onDismiss = onDismiss
) { data ->
    if (data == null) {
        return@ModalSheet
    }
    // Use `data` for content
}
```

https://user-images.githubusercontent.com/20944869/167252435-6a31b477-d1a1-40bb-b6fb-9c14f65aee05.mp4

### CODE REVIEW

**Comment by wooodenleg** [comment]
I think this makes it to have duplicit information. What is the use case to have separate `isVisible` and `data` parameters. I imagined API like this:
```kotlin
ModalSheet(
    data: T?
    content: @Composable (T) -> Unit
) { /* ... */ }
```
That means that we can close the sheet while displaying the old "data". Nullability becomes the visibility

File: modalsheet/src/main/java/eu/wewox/modalsheet/ModalSheet.kt [hunk 1]
```diff
@@ -68,15 +69,61 @@ fun ModalSheet(
     },
     content: @Composable ModalSheetScope.() -> Unit,
 ) {
+    ModalSheet(
+        data = Unit,
+        visible = { visible },
+        onDismiss = onDismiss,
```
Comment (wooodenleg) [inline modalsheet/src/main/java/eu/wewox/modalsheet/ModalSheet.kt:75]: Implementation-wise I think that "data sheet" should depend on the regular "boolean visible sheet". Not the other way around

**Comment by wooodenleg** [comment]
I belive MVP could be as easy as this:
```kotlin
@Composable
fun <T> ModalSheet(
    data: T?,
    onDismiss: () -> Unit,
    content: @Composable (T) -> Unit,
) {
    var lastNonNullData by remember { mutableStateOf(data) }
    DisposableEffect(data) {
        if (data != null) lastNonNullData = data
        onDispose {}
    }

    ModalSheet(
        visible = data != null,
        onDismiss = onDismiss
    ) {
        lastNonNullData?.let {
            content(it)
        }
    }
}
```

**Comment by oleksandrbalan** [comment]
> The problem is how to behave when the transition is not T -> null but T -> T'

Then there will be no animation 🤷‍♂️ 
At least for now

https://user-images.githubusercontent.com/20944869/167285578-32c62a04-1d72-4194-91e9-e9fefa7a22cf.mp4

**Comment by wooodenleg** [comment]
> Then there will be no animation 🤷‍♂️
At least for now

Yeah, I would love to have something like `AnimatedContent` in the future which would hide the sheet and show the new data. But I would agree that's and edge case and should be an future improvement rather than main feature

File: demo/build.gradle [hunk 2]
```diff
@@ -31,6 +31,7 @@ android {
     kotlinOptions {
         jvmTarget = "1.8"
         freeCompilerArgs = freeCompilerArgs +
+                "-Xopt-in=kotlin.RequiresOptIn" +
```
Comment (wooodenleg) [inline demo/build.gradle:34]: Does the demo module need this?

**Review by wooodenleg — APPROVED** [review]
❤️

File: demo/build.gradle (hunk 2 above)
Comment (oleksandrbalan) [reply demo/build.gradle:34]: Yeap, there are file-wide opt-ins in examples

**Pull request merged**